src/kube-runtime/GOPATH/src/github.com/microsoft/runtime/pkg/aggregator/error_aggregator.go

```diff
@@ -0,0 +1,461 @@
+	data, err := yaml.Marshal(runtimeExitInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	exitInfoSize := len(data)
+	leftSize := a.maxAggregateLogSize
+	if exitInfoSize <= leftSize {
+		return data, nil
+	}
+	remainTruncateSize := exitInfoSize - leftSize
+
+	if runtimeExitInfo.ErrorLogs != nil {
+		// truncate runtime log first
+		truncatedRuntimeLog, trucatedSize := a.truncateLog(runtimeExitInfo.ErrorLogs.Platform, remainTruncateSize)
+		runtimeExitInfo.ErrorLogs.Platform = truncatedRuntimeLog
+		remainTruncateSize = remainTruncateSize - trucatedSize
+		if remainTruncateSize <= 0 {
+			data, err := yaml.Marshal(runtimeExitInfo)
+			return data, err
+		}
+
+		// truncate the user log
+		truncatedUserLog, trucatedSize := a.truncateLog(runtimeExitInfo.ErrorLogs.User, remainTruncateSize)
+		runtimeExitInfo.ErrorLogs.User = truncatedUserLog
+		remainTruncateSize = remainTruncateSize - trucatedSize
+
+		if remainTruncateSize <= 0 {
+			data, err := yaml.Marshal(runtimeExitInfo)
+			return data, err
+		}
+	}
+
+	if runtimeExitInfo.MatchedPlatformLogString != nil {
+		// truncate the match log
+		l := len(*runtimeExitInfo.MatchedPlatformLogString) - remainTruncateSize
+		if l >= 0 {
+			runtimeExitInfo.MatchedPlatformLogString = ptrString((*runtimeExitInfo.MatchedPlatformLogString)[:l])
+			data, err := yaml.Marshal(runtimeExitInfo)
+			return data, err
+		}
+		remainTruncateSize = remainTruncateSize - len(*runtimeExitInfo.MatchedPlatformLogString)
+		runtimeExitInfo.MatchedPlatformLogString = nil
+	}
+
+	if runtimeExitInfo.MatchedUserLogString != nil {
+		l := len(*runtimeExitInfo.MatchedUserLogString) - remainTruncateSize
+		if l >= 0 {
+			runtimeExitInfo.MatchedUserLogString = ptrString((*runtimeExitInfo.MatchedUserLogString)[:l])
+			data, err := yaml.Marshal(runtimeExitInfo)
+			return data, err
+		}
+	}
+
+	return nil, errors.New("Failde to truncate the exit info")
+}
+
+// SetMaxAggregateLogSize to set maxAggregateLogSize and used for test
+func (a *ErrorAggregator) SetMaxAggregateLogSize(size int) {
+	a.maxAggregateLogSize = size
+}
+
+// ExitInfoPrefix get aggregate log prefix
+func (a *ErrorAggregator) ExitInfoPrefix() string {
+	return a.aggExitInfoBegin
+}
+
+// ExitInfoSuffix get aggregate log suffix
+func (a *ErrorAggregator) ExitInfoSuffix() string {
+	return a.aggExitInfoEnd
+}
+
+// NewErrorAggregator create an error aggregator
+func NewErrorAggregator(l *LogFiles, logger *logger.Logger) (*ErrorAggregator, error) {
+	if len(l.UserLog) == 0 || len(l.RuntimeErrorLog) == 0 {
+		return nil, errors.New("invalide log file")
+	}
+
+	if logger == nil {
+		return nil, errors.New("logger not provide")
+	}
+
+	a := ErrorAggregator{
+		logFiles:            l,
+		logger:              logger,
+		maxAggregateLogSize: 4096,
+		maxMatchLogLen:      2048,
+		maxUserLogLines:     20,
+		maxRuntimeLogLines:  10,
+		defaulExitCode:      255,
+		maxSearchLogSize:    100 * 1024 * 1024, // 100MB
+		aggExitInfoBegin:    "[PAI_RUNTIME_ERROR_START]",
```

Comment:
Can you test the worst time that exithandler need?

Reply:
It depeneds on many conditions. The log size, the  complexity of the pattern, the number of the pattern. Currently, the max number of patterns is 32 220-255 excludes 250,251,252,255. The worse case will happens when all these patterns with same exit code, and complex regex pattern. 
Can do a test with useSize larger than 100MB, all patterns with the same exit code. And use simple regex pattern such as: `command not found` for the test

---
In reply to: [322047160](https://github.com/microsoft/pai/pull/3585#discussion_r322047160) [](ancestors = 322047160)

Reply:
yep, pls benchmark it, make sure the default setting is not too time consuming

---
In reply to: [322054168](https://github.com/microsoft/pai/pull/3585#discussion_r322054168) [](ancestors = 322054168,322047160)

Reply:
Seems 100MB may also introduce false alarm, limit to 10MB?

---
In reply to: [322054707](https://github.com/microsoft/pai/pull/3585#discussion_r322054707) [](ancestors = 322054707,322054168,322047160)

Reply:
Will change to 10MB. Not do performance test yet

---
In reply to: [323139216](https://github.com/microsoft/pai/pull/3585#discussion_r323139216) [](ancestors = 323139216,322054707,322054168,322047160)